### PR TITLE
fix: [DEV] optimise notification query [#1432]

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Querying from MessagePreview can be slow

### Solutions

1. reduce complexity by add a new column is_quoting_self by to the MessageTextContent
2. get notification its query instead of using MessagePreview.

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/notification/LocalNotification.kt [hunk 1]
```diff
@@ -45,29 +50,29 @@ sealed class LocalNotificationMessage(open val author: LocalNotificationMessageA
...
+        override val time: Instant
     ) : LocalNotificationMessage(author, time)
 }
 
-data class LocalNotificationMessageAuthor(val name: String, val imageUri: String?)
+data class LocalNotificationMessageAuthor(val name: String, val imageUri: UserAssetId?)
 
 enum class LocalNotificationCommentType {
-    PICTURE, FILE, REACTION, MISSED_CALL, NOT_SUPPORTED_YET
+    PICTURE, FILE, REACTION, MISSED_CALL, KNOCK, NOT_SUPPORTED_YET
```
Comment (MohamadJaara) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/notification/LocalNotification.kt:77]: not sure why KNOCK  is missing in DEV, maybe @borichellow can help me figure out if i am missing something here

File: persistence/src/commonMain/db_user/migrations/26.sqm [hunk 2]
```diff
@@ -1,31 +1,105 @@
...
+    QuotedMessage.content_type AS quotedMessageContentType,
+    QuotedTextContent.text_body AS quotedTextBody,
+    QuotedAssetContent.asset_mime_type AS quotedAssetMimeType,
+    QuotedAssetContent.asset_name AS quotedAssetName
+FROM MessageTextContent
+LEFT JOIN Message AS QuotedMessage ON QuotedMessage.id = MessageTextContent.quoted_message_id AND QuotedMessage.conversation_id = MessageTextContent.conversation_id
+INNER JOIN User AS QuotedSender ON QuotedMessage.sender_user_id = QuotedSender.qualified_id
+INNER JOIN SelfUser AS SelfUser
+LEFT JOIN MessageTextContent AS QuotedTextContent ON QuotedTextContent.message_id = QuotedMessage.id AND QuotedMessage.conversation_id = MessageTextContent.conversation_id
+LEFT JOIN MessageAssetContent AS QuotedAssetContent ON QuotedAssetContent.message_id = QuotedMessage.id AND QuotedMessage.conversation_id = MessageTextContent.conversation_id;
```
Comment (ohassine) [inline persistence/src/commonMain/db_user/migrations/26.sqm:102]: We should not update already existing migration..This will break the process and we could lost some data

**Review by ohassine — CHANGES_REQUESTED** [review]

File: persistence/src/commonMain/db_user/migrations/26.sqm (hunk 2 above)
Comment (MohamadJaara) [reply persistence/src/commonMain/db_user/migrations/26.sqm:102]: i moved the migration files around
26 from release should stay as it is
and i considered any changes after as new (from the prod app point of view it is new)
and what was 26 in dev is is now 27

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/notification/LocalNotification.kt (hunk 1 above)
Comment (Garzas) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/notification/LocalNotification.kt:77]: Because it's connected with ping feature which is not in rc, I prepared another class to easier handle different type of notification with different sound

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageMapper.kt [hunk 3]
```diff
@@ -176,53 +175,53 @@ class MessageMapperImpl(
...
+            MessageEntity.ContentType.ASSET -> {
+                val type = message.assetMimeType?.contains("image/")?.let {
+                    if (it) LocalNotificationCommentType.PICTURE else LocalNotificationCommentType.FILE
+                } ?: LocalNotificationCommentType.FILE
+
+                LocalNotificationMessage.Comment(sender, message.date, type)
             }
 
-            is MessagePreviewEntityContent.MissedCall ->
+            MessageEntity.ContentType.KNOCK -> {
```
Comment (Garzas) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageMapper.kt:198]: Revert this change 😅 and remove `LocalNotificationCommentType.KNOCK`

File: persistence/src/commonMain/db_user/migrations/26.sqm [hunk 4]
```diff
@@ -1,31 +1,105 @@
...
+    MemberChangeContent.member_change_list AS memberChangeList,
+    MemberChangeContent.member_change_type AS memberChangeType,
+    ConversationNameChangedContent.conversation_name AS updateConversationName,
+    Conversation.name AS conversationName,
+    (Mention.user_id IS NOT NULL) AS isMentioningSelfUser,
+    TextContent.is_quoting_self AS isQuotingSelfUser,
+    TextContent.text_body AS text,
+    AssetContent.asset_mime_type AS assetMimeType,
+    (Message.creation_date > Conversation.last_read_date) AS isUnread,
+    IFNULL((Message.creation_date > IFNULL(Conversation.last_notified_date, 0)), FALSE) AS shouldNotify,
```
Comment (Garzas) [inline persistence/src/commonMain/db_user/migrations/26.sqm:67]: do we need this if NotificationMessage has it's own query?

File: persistence/src/commonMain/db_user/migrations/27.sqm [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    MemberChangeContent.member_change_list AS memberChangeList,
+    MemberChangeContent.member_change_type AS memberChangeType,
+    ConversationNameChangedContent.conversation_name AS updateConversationName,
+    Conversation.name AS conversationName,
+    (Mention.user_id IS NOT NULL) AS isMentioningSelfUser,
+    TextContent.is_quoting_self AS isQuotingSelfUser,
+    TextContent.text_body AS text,
+    AssetContent.asset_mime_type AS assetMimeType,
+    (Message.creation_date > Conversation.last_read_date) AS isUnread,
+    IFNULL((Message.creation_date > IFNULL(Conversation.last_notified_date, 0)), FALSE) AS shouldNotify,
```
Comment (Garzas) [inline persistence/src/commonMain/db_user/migrations/27.sqm:166]: same here

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageMapper.kt (hunk 3 above)
Comment (MohamadJaara) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageMapper.kt:198]: Removing LocalNotificationCommentType.KNOCK may be problematic since the enum type is already in released apps and looking at the enum column adapter we see that is uses the enum name

```kotlin
class EnumColumnAdapter<T : Enum<T>> @PublishedApi internal constructor(
  private val enumValues: Array<out T>,
) : ColumnAdapter<T, String> {
  override fun decode(databaseValue: String): T = enumValues.first { it.name == databaseValue } // this line will throw NoSuchElementException if the database have a value that does not exists in the enum type

  override fun encode(value: T) = value.name
}
```

File: persistence/src/commonMain/db_user/migrations/26.sqm (hunk 4 above)
Comment (MohamadJaara) [reply persistence/src/commonMain/db_user/migrations/26.sqm:67]: i left is just so 26.sqm matches on release and dev; otherwise we will get nasty errors when running migration

**Review by Garzas — APPROVED** [review]
✅
